src/dns_data.py

```diff
@@ -64,50 +50,36 @@ def create_dns_record_provider_data(
     return DNSRecordProviderData(dns_entries=statuses)
 
 
-def has_a_zone_changed(
+def has_changed(
     relation_data: list[tuple[DNSRecordRequirerData, DNSRecordProviderData]],
     topology: models.Topology | None,
+    last_valid_state: dict[str, typing.Any],
 ) -> bool:
-    """Check if a zone definition has changed.
+    """Check if the dns data has changed.
 
-    The zone definition is checked from a custom hash constructed from its DNS records.
+    This could be a change in a zone, or a removal/addition of a zone,
+    or a change in the topology.
+    We use the state.json file to compare the state when
+    the last configuration was minted to the current one.
 
     Args:
         relation_data: input relation data
         topology: Topology of the current deployment
+        last_valid_state: The last valid state, deserialized from a state.json file
 
     Returns:
         True if a zone has changed, False otherwise.
     """
     zones = dns_record_relations_data_to_zones(relation_data)
-    logger.debug("Zones: %s", [z.domain for z in zones])
-    for zone in zones:
-        try:
-            zonefile_content = pathlib.Path(
-                constants.DNS_CONFIG_DIR, f"db.{zone.domain}"
-            ).read_text(encoding="utf-8")
-            metadata = _get_zonefile_metadata(zonefile_content)
-        except (
-            InvalidZoneFileMetadataError,
-            EmptyZoneFileMetadataError,
-            FileNotFoundError,
-        ):
-            return True
-        if "HASH" in metadata and hash(zone) != int(metadata["HASH"]):
-            logger.debug("Config hash has changed !")
-            return True
-
-    if topology is not None and topology.is_current_unit_active:
-        try:
-            named_conf_content = pathlib.Path(
-                constants.DNS_CONFIG_DIR, "named.conf.local"
-            ).read_text(encoding="utf-8")
-        except FileNotFoundError:
-            return True
-        return (
-            _get_secondaries_ip_from_conf(named_conf_content).sort()
-            != [str(ip) for ip in topology.standby_units_ip].sort()
-        )
+
+    # We assume "zones" and "topology" keys exist in last_valid_state
+    # As it was created from load_state()
+
+    if zones != last_valid_state["zones"]:
+        return True
+
+    if topology != last_valid_state["topology"]:
+        return True
```

Comment:
A very nitpicky comment - i think your approach reads better though but i'll leave the thought.

```suggestion
    return zones != last_valid_state["zones"] or topology != last_valid_state["topology"]
```

Reply:
I actually like yours better, I'll change that in an upcoming PR